ciao/arch/architecture.md

```diff
@@ -0,0 +1,423 @@
+initiating from this instance flows from the bridge through the GRE tunnel into
+the CNCI's subnet gateway.  The IP packets are NATed as they flow through the CNCI's
+network stack and out of its network interface.  Upon reaching the cloud
+application they appear to have originated from the CNCI itself.  Instance
+172.16.0.3 can initiate external connections but is itself invisible to the
+outside world.
+
+Instance 172.16.0.2 is a little different.  It has been assigned an external IP
+address of 198.51.100.2. When an instance is assigned an external IP address two
+modifications are made to the CNCI to which that instance is connected:
+
+1. DNAT and SNAT iptables rules are set up on the CNCI to translate between the
+instance's internal and external IP address, in this case between 172.16.0.2 and
+198.51.100.2.
+2. The external IP address is added to the network interface of the CNCI making the
+IP address visible on the external network.
+
+When an external client tries to connect to 198.51.100.2 its network packets
+arrive at the network interface of the CNCI, with which the external IP address
+198.51.100.2 is associated.  The destination address of the IP packets is then
+translated by an IP table rule from 198.51.100.2 to 172.16.0.2.  The traffic
+then makes its way through the subnet gateway and the GRE tunnel to the instance
+running on the compute node.  The reverse process happens when an instance with
+an external IP address tries to connect to an external device.  The packets flow
+through the tunnel, the subnet gateway and the CNCI's network card to the outside
+world.  The source address of these packets is rewritten from the internal
+address to the external IP address, i.e., from 172.16.0.2 to 198.51.100.2, and
+not the IP address of the CNCI, 198.51.100.77.
+
+### Docker
+
+A nice feature of the Cloud Integrated Advanced Orchestrator's SDN is that
+containers and VMs can communicate with each other on their private tenant
+networks.  This inter-instance-type communication comes for free.  No special
+configuration is required by the user or the cluster administrator.  An example
+of container and VM networking can be seen in the [tutorial](/tutorial.html).
+
+Seamless container and VM networking is achieved as follows.  Containers, like
+VMs, are assigned to a subnet and allocated an IP address from that subnet.  A
+separate docker network is created on each compute node for each subnet that is
+used by a container running on that node.  Before launcher creates a new
+container on a compute node it checks to see if the docker network corresponding
+to that subnet already exists.  If it doesn't, launcher creates the network.
+The container is itself associated with this network by launcher when it is created.
+
+All of these docker subnet networks use a custom network driver called ciao.
+The ciao network driver is hosted by launcher and implements two docker network
```

Comment:
(i.e. the launcher acts as docker network plugin and docker ipam plugin)

Reply:
I've clarified this, thanks.